### PR TITLE
First phase of addressing HAVING subquery issues

### LINKED ISSUE #3433: Some subqueries in HAVING is not recursively planned
We could potentially force all subqueries in HAVING to be recursively planned via `RecursivelyPlanAllSubqueries((Node *) query->havingQual, context);` somewhere in the code.

```SQL
CREATE TABLE table_1 (key int, value int);
SELECT create_distributed_table('table_1', 'key');

CREATE TABLE table_2 (key int, value int);
SELECT create_distributed_table('table_2', 'key');
```

The following query is not very meaningful, as we'd mostly expect many results in the subquery

```SQL
SELECT count(*) FROM table_1 GROUP BY key HAVING sum(value) > (SELECT sum(value) FROM table_2 GROUP BY key);
ERROR:  complex joins are only supported when all distributed tables are co-located and joined on their distribution columns
```

However, the following might be a useful query, still erroring out the same:

```SQL
SELECT count(*) FROM table_1 GROUP BY key HAVING avg(value) > avg((SELECT sum(value) FROM table_2 GROUP BY key));
ERROR:  complex joins are only supported when all distributed tables are co-located and joined on their distribution columns
```

The current workaound is to add `OFFSET 0` to the subquery, which hits #3432

```SQL
SELECT count(*) FROM table_1 GROUP BY key HAVING avg(value) > avg((SELECT sum(value) FROM table_2 GROUP BY key OFFSET 0));
ERROR:  cannot handle unplanned sub-select
```

### PR DESCRIPTION
**Pull request by serprex**
Add failing tests, make changes to avoid crashes at least

Fix HAVING subquery pushdown ignoring reference table only subqueries, also include HAVING in recursive planning

Given that we have a function IsDistributedTable which includes reference tables, it seems best to have IsDistributedTableRTE & QueryContainsDistributedTableRTE reflect that they do not include reference tables in their check

Similarly SublinkList's name should reflect that it only scans WHERE

contain_agg_clause asserts that we don't have SubLinks, use contain_aggs_of_level as suggested by pg sourcecode

This causes some previous test cases which gave results to raise errors instead

DESCRIPTION: improve HAVING subquery planning correctness

This seeks to be a conservative version of #3489 suitable for backporting
Fixes #3433

### CODE REVIEW

File: src/test/regress/expected/multi_reference_table.out [hunk 1]
```diff
@@ -1602,6 +1601,48 @@ BEGIN;
 ALTER TABLE reference_table_test ADD COLUMN value_dummy INT;
 INSERT INTO reference_table_test VALUES (2, 2.0, '2', '2016-12-02');
 ROLLBACK;
+-- Previous issue failed to rename reference tables in subqueries
+EXPLAIN SELECT value_1, count(*) FROM colocated_table_test GROUP BY value_1
```
Comment (marcocitus) [inline src/test/regress/expected/multi_reference_table.out]: Should turn off costs

**Comment by serprex** [comment]
```
 SELECT value_1, count(*) FROM colocated_table_test GROUP BY value_1
 HAVING (SELECT rt.value_2 FROM reference_table_test rt where rt.value_2 = 2) > 0
 ORDER BY 1;
- value_1 | count
----------------------------------------------------------------------
-       1 |     4
-       2 |     5
-(2 rows)
-
+ERROR:  relation "public.colocated_table_test_1250005" does not exist
```

Very odd. Not happening on my own system

**Comment by serprex** [comment]
@onderkalaci could I get your review on the changes to `local_executor.c`?

File: src/backend/distributed/planner/recursive_planning.c [hunk 2]
```diff
@@ -314,6 +315,25 @@ RecursivelyPlanSubqueriesAndCTEs(Query *query, RecursivePlanningContext *context
...
+		List *havingSublinks = NIL;
+		ExtractSublinkWalker(query->havingQual, &havingSublinks);
+
+		SubLink *havingSublink;
+		foreach_ptr(havingSublink, havingSublinks)
+		{
+			if (ContainsReferencesToOuterQuery(castNode(Query, havingSublink->subselect)))
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "Cannot recursively plan HAVING with references to outer query",
```
Comment (marcocitus) [inline src/backend/distributed/planner/recursive_planning.c]: This probably wouldn't be clear to users unfamiliar with recursive planning. Maybe "Subqueries in HAVING cannot refer to the outer query".

File: src/backend/distributed/executor/local_executor.c [hunk 3]
```diff
@@ -319,18 +319,20 @@ ExtractLocalAndRemoteTasks(bool readOnly, List *taskList, List **localTaskList,
 			 * nodes. We always prefer to use local placement, and require remote
 			 * placements only for modifications.
 			 */
-			task->partiallyLocalOrRemote = true;
 
-			Task *localTask = copyObject(task);
+			/* Don't push a task without any placements */
```
Comment (pykello) [inline src/backend/distributed/executor/local_executor.c]: This comment confuses me. Few lines ago we have `Assert(!(localTaskPlacementList == NIL && remoteTaskPlacementList == NIL));`, so we shouldn't have tasks without any placements here.

If we can have tasks without any placements, then why not change the assert with `if (localTaskPlacementList == NIL && remoteTaskPlacementList == NIL) continue;`?

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 4]
```diff
@@ -402,48 +404,66 @@ FindNodeCheckInRangeTableList(List *rtable, bool (*check)(Node *))
...
 /*
- * QueryContainsDistributedTableRTE determines whether the given
+ * QueryContainsDistributedNonReferenceTableRTE determines whether the given
  * query contains a distributed table.
  */
 bool
-QueryContainsDistributedTableRTE(Query *query)
+QueryContainsDistributedNonReferenceTableRTE(Query *query)
 {
 	return FindNodeCheck((Node *) query, IsDistributedTableRTE);
```
Comment (pykello) [inline src/backend/distributed/planner/multi_logical_planner.c]: IsDistributedNonReferenceTableRTE?

Comment (pykello) [inline src/backend/distributed/planner/multi_logical_planner.c]: Did you intend IsDistributedNonReferenceTableRTE?

File: src/backend/distributed/planner/recursive_planning.c [hunk 5]
```diff
@@ -1025,7 +1044,7 @@ RecursivelyPlanSetOperations(Query *query, Node *node,
 		Query *subquery = rangeTableEntry->subquery;
 
 		if (rangeTableEntry->rtekind == RTE_SUBQUERY &&
-			QueryContainsDistributedTableRTE(subquery))
+			QueryContainsDistributedNonReferenceTableRTE(subquery))
```
Comment (pykello) [inline src/backend/distributed/planner/recursive_planning.c]: Add comment why this is ok.

File: src/backend/distributed/planner/query_colocation_checker.c [hunk 6]
```diff
@@ -195,7 +195,7 @@ SubqueryColocated(Query *subquery, ColocatedJoinChecker *checker)
 	 */
 	if (list_length(filteredRestrictionList) == 0)
 	{
-		Assert(!QueryContainsDistributedTableRTE(subquery));
+		Assert(!QueryContainsDistributedNonReferenceTableRTE(subquery));
```
Comment (pykello) [inline src/backend/distributed/planner/query_colocation_checker.c]: Previous version seemed to match the comment above?

File: src/backend/distributed/planner/query_pushdown_planning.c [hunk 7]
```diff
@@ -730,7 +730,8 @@ FromClauseRecurringTupleType(Query *queryTree)
 		return RECURRING_TUPLES_EMPTY_JOIN_TREE;
 	}
 
-	if (FindNodeCheckInRangeTableList(queryTree->rtable, IsDistributedTableRTE))
+	if (FindNodeCheckInRangeTableList(queryTree->rtable,
+									  IsDistributedNonReferenceTableRTE))
```
Comment (pykello) [inline src/backend/distributed/planner/query_pushdown_planning.c]: Do we have a test to validate this change? If not, please add one.

**Comment by pykello** [comment]
Other than the comments I added, this seemed fine. I want to do some more testing and have a second look tomorrow before approving it though.

File: src/test/regress/expected/multi_subquery.out [hunk 8]
```diff
@@ -884,8 +884,165 @@ AS foo;
...
+SELECT count(*) FROM keyval1 GROUP BY key HAVING sum(value) > (SELECT sum(value) FROM keyvalref GROUP BY key);
+                                 QUERY PLAN
+---------------------------------------------------------------------
+ HashAggregate
+   Group Key: remote_scan.worker_column_2
+   Filter: ((pg_catalog.sum(remote_scan.worker_column_3))::bigint > $0)
+   InitPlan 1 (returns $0)
+     ->  Function Scan on read_intermediate_result intermediate_result
+   ->  Custom Scan (Citus Adaptive)
+         ->  Distributed Subplan XXX_1
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_subquery.out]: Do we need the plans on the workers? If not maybe use `public.coordinator_plan`

File: src/backend/distributed/planner/query_pushdown_planning.c (hunk 7 above)
Comment (serprex) [reply src/backend/distributed/planner/query_pushdown_planning.c]: This isn't changed. I renamed `IsDistributedTableRTE` to `IsDistributedNonReferenceTableRTE` since that's what it actually was implemented as. grep for `IsDistributedTableRTE` for which cases were truly updated

File: src/backend/distributed/planner/multi_logical_planner.c (hunk 4 above)
Comment (serprex) [reply src/backend/distributed/planner/multi_logical_planner.c]: Yes

File: src/backend/distributed/planner/query_pushdown_planning.c (hunk 7 above)
Comment (marcocitus) [reply src/backend/distributed/planner/query_pushdown_planning.c]: maybe
DistributedNonReferenceTable -> DistributedTable
DistributedTable -> CitusTable

File: src/backend/distributed/executor/local_executor.c (hunk 3 above)
Comment (serprex) [reply src/backend/distributed/executor/local_executor.c]: It's possible here for `remoteTaskPlacementList != NIL`

File: src/backend/distributed/planner/query_pushdown_planning.c (hunk 7 above)
Comment (serprex) [reply src/backend/distributed/planner/query_pushdown_planning.c]: Still confusing unless we rename `IsDistributedTable` to `IsCitusTable`

Comment (marcocitus) [reply src/backend/distributed/planner/query_pushdown_planning.c]: That's what I meant

File: src/backend/distributed/planner/recursive_planning.c (hunk 5 above)
Comment (serprex) [reply src/backend/distributed/planner/recursive_planning.c]: Not ok, updated

File: src/backend/distributed/executor/local_executor.c (hunk 3 above)
Comment (pykello) [reply src/backend/distributed/executor/local_executor.c]: 0. What happens when a task doesn't have any local placements and is read only? Shouldn't we add it to *remoteTaskList?
1. The comment is still confusing. According to the assert, we cannot have "task without any placements", you might want to reword it to "task without any local placements"
2. The comment before this might need this updating. Intermediate results might not have both local and remote nodes.

Comment (serprex) [reply src/backend/distributed/executor/local_executor.c]: https://github.com/citusdata/citus/blob/master/src/backend/distributed/executor/adaptive_executor.c#L712

The current case is leaving localTaskList empty, so `tasksToExecute` comes out as it should be. It seems the comments are inaccurate, we claim `tasksToExecute` is the union of `localTaskList` & `remoteTaskList`, meanwhile this code path drops remote tasks

File: src/test/regress/expected/multi_subquery.out (hunk 8 above)
Comment (serprex) [reply src/test/regress/expected/multi_subquery.out]: The key indicator of what's being tested is `Seq Scan on reference_table_test_1250000 rt`

File: src/backend/distributed/planner/multi_logical_optimizer.c [hunk 9]
```diff
@@ -1434,7 +1435,7 @@ MasterExtendedOpNode(MultiExtendedOp *originalOpNode,
 		Expr *originalExpression = originalTargetEntry->expr;
 		Expr *newExpression = NULL;
 
-		bool hasAggregates = contain_agg_clause((Node *) originalExpression);
+		bool hasAggregates = contain_aggs_of_level((Node *) originalExpression, 0);
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_logical_optimizer.c]: These changes are probably somewhat obsolete after the recursive planning change. Reverting them does not yield test failures. 

I wonder whether it makes sense to keep the old call with the Assert since we really don't know how to handle Sublinks here, but either way works for me.

File: src/backend/distributed/planner/recursive_planning.c [hunk 10]
```diff
@@ -1302,6 +1315,29 @@ ContainsReferencesToOuterQueryWalker(Node *node, VarLevelsUpWalkerContext *conte
...
 
+/*
+ * NodeContainsSubqueryReferencingOuterQuery determines whether the given node
+ * contains anything that points outside of the query itself.
+ */
+static bool
+NodeContainsSubqueryReferencingOuterQuery(Node *node)
+{
+	List *sublinks = NIL;
+	ExtractSublinkWalker(node, &sublinks);
```
Comment (marcocitus) [inline src/backend/distributed/planner/recursive_planning.c]: I was concerned about this function not descending into RTE_SUBQUERY, but since the caller does recurse it still catches everything.

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 9 above)
Comment (serprex) [reply src/backend/distributed/planner/multi_logical_optimizer.c]: Not included in this PR was not pushing down sublink arguments to aggregates. A test is included in this PR for `avg((subquery))`, but the expected result is `ERROR:  cannot handle unplanned sub-select`. This is being raised by `get_agg_clause_costs`. The rest of the fix will be deferring calls to `get_agg_clause_costs` until we process sublinks & not allowing sublinks to be pushed down. The sublinks after recursive planning are subqueries reading intermediate results, which obviously can't be pushed down

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 11]
```diff
@@ -412,37 +413,55 @@ QueryContainsDistributedTableRTE(Query *query)
...
+ * IsDistributedTableRTE gets a node and returns true if the node
+ * is a range table relation entry that points to a distributed relation,
+ * returning false still if the relation is a reference table.
+ */
+bool
+IsDistributedTableRTE(Node *node)
+{
+	Oid relationId = NodeTryGetRteRelid(node);
+	return relationId != InvalidOid && IsCitusTable(relationId) &&
+		   PartitionMethod(relationId) != DISTRIBUTE_BY_NONE;
```
[the remark this answers is not in the document]
Comment (serprex) [reply src/backend/distributed/planner/multi_logical_planner.c]: Added to list of things to batch in a future cleanup PR. Would like to give other branches more time to rebase before reintroducing `IsDistributedTable` with same signature but different semantics

File: src/backend/distributed/planner/multi_logical_optimizer.c (hunk 9 above)
Comment (marcocitus) [reply src/backend/distributed/planner/multi_logical_optimizer.c]: Actually, intermediate results are typically broadcast to all nodes, so subqueries that read from them can be pushed down, assuming that does not create a correctness problem.

**Comment by serprex** [comment]
This doesn't apply cleanly to release-9.2. Will create a PR for cherry picking this commit to release-9.2
